File: paimon-core/src/main/java/org/apache/paimon/consumer/Consumer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.consumer;

import org.apache.paimon.fs.FileIO;
import org.apache.paimon.fs.Path;
import org.apache.paimon.utils.JsonSerdeUtil;

import org.apache.paimon.shade.jackson2.com.fasterxml.jackson.annotation.JsonCreator;
import org.apache.paimon.shade.jackson2.com.fasterxml.jackson.annotation.JsonGetter;
import org.apache.paimon.shade.jackson2.com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import org.apache.paimon.shade.jackson2.com.fasterxml.jackson.annotation.JsonProperty;
import org.apache.paimon.shade.jackson2.com.fasterxml.jackson.databind.exc.MismatchedInputException;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.Optional;

/** Consumer which contains next snapshot. */
@JsonIgnoreProperties(ignoreUnknown = true)
public class Consumer {

    private static final String FIELD_NEXT_SNAPSHOT = "nextSnapshot";

    private final long nextSnapshot;

    @JsonCreator
    public Consumer(@JsonProperty(FIELD_NEXT_SNAPSHOT) long nextSnapshot) {
        this.nextSnapshot = nextSnapshot;
    }

    @JsonGetter(FIELD_NEXT_SNAPSHOT)
    public long nextSnapshot() {
        return nextSnapshot;
    }

    public String toJson() {
        return JsonSerdeUtil.toJson(this);
    }

    public static Consumer fromJson(String json) {
        return JsonSerdeUtil.fromJson(json, Consumer.class);
    }

    public static Optional<Consumer> fromPath(FileIO fileIO, Path path) {
        int retryNumber = 0;
        MismatchedInputException exception = null;
        while (retryNumber++ < 5) {
            try {
                return fileIO.readOverwrittenFileUtf8(path).map(Consumer::fromJson);
            } catch (MismatchedInputException e) {
                // retry
                exception = e;
                try {
                    Thread.sleep(100);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    throw new RuntimeException(ie);
                }
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        }
        throw new UncheckedIOException(exception);
    }
}

File: paimon-python/pypaimon/consumer/consumer.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################
"""Consumer dataclass for streaming read progress."""

import json
import time
from dataclasses import dataclass
from typing import Optional


@dataclass
class Consumer:
    """Consumer which contains the next snapshot to be read."""

    next_snapshot: int

    def to_json(self) -> str:
        """Serialize to JSON."""
        return json.dumps({"nextSnapshot": self.next_snapshot})

    @staticmethod
    def from_json(json_str: str) -> 'Consumer':
        """Deserialize from JSON."""
        data = json.loads(json_str)
        return Consumer(next_snapshot=data["nextSnapshot"])

    @classmethod
    def from_path(cls, file_io, path: str) -> Optional['Consumer']:
        """
        Load consumer from file path with retry mechanism.

        Args:
            file_io: FileIO instance for reading files
            path: Path to consumer file

        Returns:
            Consumer instance if file exists, None otherwise

        Raises:
            RuntimeError: If retry fails after 10 attempts
        """
        retry_number = 0
        exception = None
        while retry_number < 10:
            try:
                content = file_io.read_file_utf8(path)
            except FileNotFoundError:
                return None
            except Exception as e:
                exception = e
                time.sleep(0.2)
                retry_number += 1
                continue
            try:
                return cls.from_json(content)
            except Exception as e:
                exception = e
                time.sleep(0.2)
                retry_number += 1
                continue
        raise RuntimeError(f"Retry fail after 10 times: {exception}")
